tests/ui/self_named_constructor.stderr

```diff
@@ -0,0 +1,32 @@
+error: constructor `q` has the same name as the type
+  --> $DIR/self_named_constructor.rs:6:5
+   |
+LL | /     pub fn q() -> Q {
+LL | |         Q
+LL | |     }
+   | |_____^
+   |
+   = note: `-D clippy::self-named-constructor` implied by `-D warnings`
+help: consider renaming `q()` to `new()`
+  --> $DIR/self_named_constructor.rs:6:5
+   |
+LL | /     pub fn q() -> Q {
+LL | |         Q
+LL | |     }
+   | |_____^
+
+error: constructor `default` has the same name as the type
```

Comment:
Is it expected behavior?

Reply:
It is the expected behavior, but I'm not sure if that's the behavior we want.

The behavior happens because `default` is a function in the trait, and the struct itself is named `Default`. I ran into this when running `cargo test`.

Reply:
I don't have a strong opinion here, but sounds like a false positive.

IMHO is it's either of:
- the trait's fault to be focused on the concrete types, or
- **not** the structure that implements the trait fault to be named the same as one of the trait methods.

Like, the implementor should not be punished for the fact that it implements a trait, even if the trait has odd naming choices.

Reply:
Also, I'd say that for traits it's more than OK to have not conventional constructor names, because conventional ones are expected to be used by structures/enums directly.

Probably, `impl XXX for YYY` blocks should be checked at all.

Reply:
Yeah traits _are_ usually named the same as their methods.

Reply:
The thing is that here the trait name doesn't really matter.
If I understand everything correctly, the following code will trigger the lint.

```rust
trait Foo {
    fn bubble() -> Self
}

impl Foo for Bubble {
    // Name is not related to the trait at all, but the structure just happened to have the same name.
    fn bubble() -> Self { Bubble { ... } }
}
```

I think it's a rare occasion to hit that accidentally, but anyway weird.
The more I think of it, the more I lean toward not analyzing trait `impl`s.

Reply:
Ah, I mean, we shouldn't be analyzing trait impls at all, yes